src/common/cockpitwebcertificate.c

```diff
@@ -111,6 +111,41 @@ cockpit_certificate_locate (char **error)
   return NULL;
 }
 
+/**
+ * cockpit_certificate_key_path:
+ *
+ * Check if there is a matching *.key file for the given *.cert file.
+ *
+ * Return key file path on success, or %NULL if there is no matching *.key file.
+ */
+char *
+cockpit_certificate_key_path (const char *certfile)
+{
+  size_t len = strlen (certfile);
+  char *keypath = NULL;
+
+  /* invalid if it does not end with .cert */
+  if (len <= 5 || strcmp (certfile + len - 5, ".cert") != 0)
+    {
+      warnx ("invalid certificate file name: %s", certfile);
+      return NULL;
+    }
+
+  keypath = strdupx (certfile);
+  keypath[len - 4] = 'k';
+  keypath[len - 3] = 'e';
+  keypath[len - 2] = 'y';
+  keypath[len - 1] = '\0';
+
+  if (access (keypath, F_OK) < 0)
```

Comment:
`access()` is weird.  It almost never makes sense to call it.

Why not have this function open and return the contents of the `.key` file instead?

Also: you should treat `ENOENT` as an expected error and all other errors as failures.

Reply:
Returning the contents instead of the path would be rather inconvenient -- in all cases where we call it, we want the path, as that's what `connection_crypto_init()` or `g_tls_certificate_new_from_files()` takes. In particular, having the cert as path and the key as value would require more manual work on our side.

I specifically used `F_OK` instead of `R_OK` to not just silently ignore key files with wrong permissions, but getting an explicit error message about them later on, when we try to open it. 

ENOENT> fair enough, I can raise a `warn()` about it, in case there's weird errors like `ELOOP` or `ENOTDIR`.

What would you suggest instead of `access()`?

Reply:
The follow-up adds the warning for not-`ENOENT`.

Reply:
I reworked this a bit harder, and changed from this LBYL to a more robust EAFP approach. I like this much better, thanks!